### PR TITLE
Bugfix: Prevent duplicate records in paginated lists and exports

### PR DESCRIPTION
**Pull request by andrewbrazzatti**
## Summary

Fix duplicate and missing records across paginated record lists and CSV/JSON exports when records share a sort value. Mongo queries now include a unique `_id` tie-breaker.

---

## Context / Motivation

A production RDMP export and API traversal returned the expected row count while omitting records. Sorting only by `lastSaveDate` allowed equal timestamps to appear in different orders between `skip`/`limit` queries. A separate total count cannot detect those duplicates and omissions.

---

### Consistent pagination

- Apply a shared sort builder to active-record lists, deleted-record lists, and exports.
- Append `_id` ascending after requested primary and secondary fields, preserving an explicit `_id` direction.
- Support JSON objects, JSON sort tuples, and `field:direction` expressions; apply bare-field expressions in descending order.
- Sort deleted records using `deletedRecordMetadata` for record fields while preserving outer `_id`, `redboxOid`, and `dateDeleted` fields.
- Map deleted-list UI columns (`title`, `dateCreatedDisplay`, `dateModifiedDisplay`, and `dateDeletedDisplay`) to their persisted fields.
- Reject JSON null, primitives, malformed sort tuples, and invalid primary or secondary directions with a client validation error before querying MongoDB. Preserve numeric directions, direction names, and MongoDB `$meta` expressions.
- Require non-empty fields and exact colon syntax: primary sorts accept `field` or `field:direction`; secondary sorts require `field:direction`. Empty segments and trailing segments return a client validation error.

### Regression coverage

- Cover sort precedence, explicit tie-breakers, and both streamed export formats.
- Add Mongo integration tests with 3,068 synthetic records whose repeated timestamps cross page boundaries, including a partial final page.
- Document how to run the integration tests and the limits of offset pagination.

---

## Testing

- Reproduced the defect against MongoDB before the fix: 3,068 returned rows contained only 1,048 unique OIDs.
- Ran the full storage package suite in Docker with `MONGO_TEST_URL` configured: **226 tests passed**.
- All 17 Mongo integration scenarios returned 3,068 unique records in the expected order across active/deleted lists and CSV/JSON exports.
- Clean local coverage: 93.05% of storage package lines, including every changed sort line.
- Storage package TypeScript build and lint passed; lint reported zero warnings and errors.
- Mongo integration tests create and remove an isolated database. They are skipped when `MONGO_TEST_URL` is absent.

---

### Shared Mongo sorting

The storage adapter now centralizes sort parsing and tie-breaking for the affected queries, replacing duplicated list-query logic.

### Concurrent changes

The fix makes equal sort values deterministic. Offset pagination still does not provide a snapshot when matching records are inserted, deleted, or updated during traversal.

---

## Backwards Compatibility

Existing query parameters, response formats, and explicit sort precedence are retained. Previously unordered ties now use `_id`. Bare-field sort expressions, which previously failed to populate the query sort, are now applied in descending order.

---

## Deployment Notes

Rebuild the Mongo storage package and deploy the updated portal image. No database migration is required. Regenerate affected exports and restart API traversals from the first page after deployment.

---

## Impact

RDMP exports and paginated lists include each matching record once for a stable dataset, including records that share modification timestamps.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved record and deleted-record pagination to maintain consistent ordering when multiple records share the same timestamp or title.
  - Updated CSV and JSON exports to use deterministic ordering across pages and export results.
  - Improved deleted-record sorting for title and date display fields.
  - Added validation for unsupported sort formats and directions.
  - Added coverage for pagination, sorting, and export ordering, including large datasets and duplicate timestamps.

- **Documentation**
  - Documented pagination ordering behavior, export sorting, and integration test requirements.

### CODE REVIEW

**Comment by andrewbrazzatti** [comment]
@greptile review

**Comment by andrewbrazzatti** [comment]
@greptile review

File: packages/sails-hook-redbox-storage-mongo/src/services/MongoStorageService.ts [hunk 1]
```diff
@@ -674,6 +675,62 @@ export namespace Services {
...
+            return 1;
+          case '-1':
+          case 'desc':
+          case 'descending':
+            return -1;
+        }
+      } else if (direction !== null && typeof direction === 'object' && !Array.isArray(direction) &&
+        '$meta' in direction && typeof direction.$meta === 'string') {
+        return { $meta: direction.$meta };
+      }
```
[the remark this answers is not in the document]
Comment (andrewbrazzatti) [reply packages/sails-hook-redbox-storage-mongo/src/services/MongoStorageService.ts:693]: Verified against f8e044862: getSortDirection returns a new `{ $meta: direction.$meta }` object, so the example’s `extra` property is removed rather than forwarded. A direct runtime check confirms the normalized result is `{ score: { $meta: "textScore" }, _id: 1 }`.

The driver’s documented SortDirection contract accepts `{ $meta: string }`: https://mongodb.github.io/node-mongodb-native/7.0/types/SortDirection.html. This validator deliberately preserves that existing format. Metadata keyword availability depends on the server, and valid keywords also have query prerequisites (for example, textScore requires a $text predicate). A hard-coded keyword allowlist here would narrow existing driver support without establishing that a query is valid.

Resolving as non-actionable for this sort-format validation change: extra properties are already normalized away, while metadata semantics remain with MongoDB.
